### ISSUE TITLE
Rubygems only sources rubygems.org

### ISSUE THREAD

**Issue by bradrobertson** [rubygems/rubygems#2731]
I would like to suggest a feature.

My current problem is that, from what I can tell, a gemspec doesn't allow you to specify different sources where gems can be installed from. Bundler allows us to do this, but I can't define a transitive dependency since a gem from my custom source defined in Bundler would need its gemspec to be able to define the source of _that_ dependency. 

This issue is related to:

  - [+] Publishing a library

I'm sure I can't be the first to mention this so I'm wondering if there's an ongoing discussion anywhere around allowing gemspecs to specify gem sources? (beyond rubygems.org)

I know there's also gemstash which acts as a rubygems.org cache, but from what I understand, if there are naming collisions, rubygems.org wins out??

**Comment by Fire-Dragon-DoL** [comment]
You are totally right @bradrobertson , I faced this issue multiple times and would love to see it addressed.

**Comment by deivid-rodriguez** [comment]
Could you explain the exact issue you're having @Fire-Dragon-DoL? I'm not sure about what the problem is.

Bundler allows you to define different sources where your gems should be picked from, and as pointed out, a gemspec can be pushed to any gem server source (for example, to a gemstash instance). So I'm not sure how coupling gemspecs with gem sources would benefit people.

**Comment by sbellware** [comment]
@deivid-rodriguez Here's an angle to consider it from:

Some producer of a package publishes their package. They publish the package on their own package repository. That package has dependencies on other packages published in that same package repository (ie: their own package repository).

The producer of the package has an intent. That intent is that the package depends on a specific package that is in a specific namespace. That namespace is their package repository.

Reduced to a rule, it would be something like, "For my package _A_ that depends on my package _B_, where both packages are published to my own repository address _R_, I expect that _R.A_ installs _R.B_.

What is currently the case with Gemspec is the rule, "For my package _A_ that depends on my package _B_, where both packages are published to my own repository address _R_, I expect that any _A_ in any repository in the entire ecosystem installs any _B_ in any repository in the entire ecosystem.

Because Gemspec doesn't allow for explicit namespacing over the address space of the ecosystem wherein packages are contained, there's no way for the publisher to specify their explicit intent.

So, we use Bundler to act as an override and a contingency. But that means that dependencies effectively have to be declared twice: Dependency names _without_ namespacing are declared in Gemspec, and Dependency names _with_ namespacing are declared in Gemfile.

The suggestion, I believe, is to account for namespacing in Gemspec, and to use a very well-known and well-adopted standard for encoding namespaces across a potentially-infinite system of names and naming: URI namespaces.

And if URI namespaces are part of a Gemspec's declaration of a dependency, then the package installation tool can choose to use that namespace as the network address of the package.

And, of course, override's can be applied _optionally_ if and only if the situation demands an override, and not as a matter of course.

The opposition to this that has been expressed in the past is that a publisher can make their package unavailable if they choose. But that's already the case today, as well. And ultimately, it's addressed by the override mechanisms already in place. And ultimately, it's the package owner's prerogative to discontinue availability of a package, anyway.

It would seem that we've gotta do things with the combination of RubyGems and Bundler that should not be necessary but for global namespace inherent in Gemspec and RubyGems currently, and that the embrace of URIs and a namespaced address space for Ruby dependencies would go a long way to clearing some of the more esoteric and cluttered aspects of Ruby package management.

**Comment by deivid-rodriguez** [comment]
> So, we use Bundler to act as an override and a contingency. But that means that dependencies effectively have to be declared twice: Dependency names without namespacing are declared in Gemspec, and Dependency names with namespacing are declared in Gemfile.

Could you clarify this? `gemspec` and `Gemfile` have different purposes. `gemspec` is used by library authors to specify the dependencies of their library. `Gemfile` is used by library users to specifiy the dependencies of their application.

In bundler, if you wrap a gem or set of gems under a source block, the gem and their dependencies should be picked up from that source, which is what you want?

**Comment by Fire-Dragon-DoL** [comment]
@deivid-rodriguez This poses essentially a lock where library authors can depend only on rubygems as package repository.

**Comment by sbellware** [comment]
@deivid-rodriguez 

> gemspec and Gemfile have different purposes

Yes, but I'm describing a world where Gemspec isn't missing the other half of its nature, and is fully-capable of standing on its own without having to be completed by another tool.

As a library author distributing code with RubyGems, I don't in fact specify the dependencies of a library. I partially specify the dependencies of a library. And then I use Bundler to turn that partial specification into one that is complete enough to fully represent my intent.

As a library author, I do this only because of the design that RubyGems and Bundler have arrived at organically over the years.

**Comment by deivid-rodriguez** [comment]
Hi again both. It's being a little hard for me to understand what's being requested here.

If it's a flag to `gem install` like `--from-source https://my.gem.server` to specifiy "I want to install this gem and its dependencies from this host", then I guess it makes sense (although it will probably take little priority in a "bundler world"). If it's adding a new field to the gemspec, then I don't get it. As explained [here](https://github.com/rubygems/rubygems/pull/218#issuecomment-2795988), I don't think it makes sense to hold this information in the gemspec file.

> As a library author distributing code with RubyGems, I don't in fact specify the dependencies of a library. I partially specify the dependencies of a library. And then I use Bundler to turn that partial specification into one that is complete enough to fully represent my intent.

This doesn't make much sense to me since `bundler` is not really the tool to specify the dependencies of a gem (`Gemfile`'s inside gems are ignored at runtime of the gem). It's end users who are responsible to choose which sources they want to pick up their gems from.

I think it'd be easier if you explain exactly how you're using rubygems & bundler as gem authors. Ideally link to some sample gem repositories and gem server instances showcasing it, and demonstrating how the current tools get in the middle for you and how you would like them to behave. Like, 'I always need to run command "X;Y;Z" but it'd be simpler if I could just run command 'A'", or "my users always need to do thing 'X', but if I could do thing 'Y', they wouldn't need to do anything and things would just work for them". The conversation so far is too abstract.

**Comment by Fire-Dragon-DoL** [comment]
Hey @deivid-rodriguez , a simple example is the following:

I'm hosting a `.gem` file on GitHub for a private gem (could even be public) and I don't want to deal with a private gem server. Since for `gem` command a `.gem` file is sufficient, installing this gem locally boils down to manually downloading the file and running `gem install mygem.gem`.
The gem is a utility library with an executable.

Now my private gem depends on another private gem, again on GitHub. This can't be achieved, the `gemspec` can't specify where the gem is located, I can't install the `.gem`, I need to specify all the dependencies in a Gemfile, even though they are (or should) be specified in gemspec.

Being able to specify a URL to a git repository containing a gemspec (or even multiple), or a git url would simplify this situation drammatically.

**Comment by indirect** [comment]
So first, I would like to acknowledge that you are correct that RubyGems does not allow you to specify a source for a gem. That was a deliberate choice on the part of the RubyGems team, because domains expire, and gemspecs should not stop working because someone forgot to renew a domain.

It is the current position of the combined RubyGems and Bundler teams that if you need to install a set of gems from a specific source, you have an “application” that should use a Gemfile to specify the list of gems and sources. Bundler and RubyGems are a single system, Bundler is included in every RubyGems release, and every Bundler feature _is_ a feature of RubyGems.

Second, I would like to point out a few ways you can install multiple private gems without running your own private gem server:

1) GitHub offers a gem server as part of hosting releases. You can simply use GitHub’s server, since you already mentioned using GitHub’s releases.

2) Bundler already supports installing multiple private gems from one or more git  repos. Use Bundler with one or more git repo URLs.

**Comment by Fire-Dragon-DoL** [comment]
To be precise, I'm talking about libraries, so Bundler is not an option.

As for the URL: shouldn't that be a prerogative of the library author?
If I want to make my library unavailable, it should be.

It's the developer taking on the dependency that will carefully evaluate and decide if the risk of a broken build is something that can be worked with.

GitHub does offer a gem server, but this is indeed limited to github, I host some of my source code on a private server and can't use it without starting a gem server.

**Comment by indirect** [comment]
I can understand that you might prefer a system like e.g. Maven, but RubyGems has chosen to (deliberately) decouple gems from servers.

If you need to install from a specific server, Gemfiles are the solution. Any git repository is a valid source for a Gemfile, so you do not need to run a gem server. 

Can you explain in more detail why Gemfiles with git repo sources are not sufficient to accomplish your library distribution goals, and why you believe adding hard coded sources to gemspecs would resolve this problem in a way that Bundler cannot?

**Comment by Fire-Dragon-DoL** [comment]
@indirect sorry, maybe I misunderstood or missed something important.

How can the Gemfile be used in the context of a library?

**Comment by indirect** [comment]
The `gem install` command installs gems locally. The `bundle install --system` command installs gems locally. They install gems into the same place, able to be used the same way (by calling `require`). Running `bundle install` on a Gemfile is the way to install specific libraries from specific sources. What do you want to do that you can't do using Bundler and a Gemfile?

**Comment by deivid-rodriguez** [comment]
I think after https://github.com/rubygems/rubygems/issues/2731#issuecomment-699537969 I finally understood the request. The user wants to do something like [gem install https://github.com/githubsvnclone/rdoc.git](https://github.com/rdp/specific_install), and automatically get all dependencies resolved to whenever the gem author specify dependencies should live in the target gemspec.

The alternative is the user writing a proper `Gemfile` and running `bundle install`.

So I guess this comes down to two requests:

1) Support something like specific_install in rubygems.
2) Allow gem authors specifying the source where their dependencies should live inside the gemspec.

Is this something along the lines of what's being requested?

I think both are quite unlikely to be implemented. but I don't strong oppose to 1).

**Comment by Fire-Dragon-DoL** [comment]
Yes, @deivid-rodriguez based on your description that would work.

**Comment by sbellware** [comment]
Just as a matter of clarification and commentary:

What I've pointed out is that Gemspec is _not_ a way for a gem to declare dependencies. 

A Gemspec is a way for a gem to declare _only the names of dependencies_.

A de facto dependency requires both a name and an address. In Ruby, resolving a name and an address requires the name to be declared in one tool, and the address to be declared in yet another tool. Unless, of course, the rubygems.org address namespace can suffice as a default. And that, ultimately, creates the challenges associated with having RubyGems effectively be a global namespace.

The ultimate design problem I'm hinting at with regard to RubyGems and Bundler is the lack of clear delineation of responsibilities between the two.

If RubyGems were _just_ a dependency loader and Bundler were _just_ a dependency installer, it would make more sense. But those lines are often blurred.

As a design, it would be desirable to say that `some_gem` depends on `some_other_gem` as such:

```
Gem::Specification.new do |s|
  s.name = 'some_gem'
  s.add_runtime_dependency 'https://my-gem-host/some_other_gem'
end
```

Or, do to it as such:

```
Gem::Specification.new do |s|
  s.name = 'some_gem'
  s.add_runtime_dependency 'some_other_gem'
end
```

As such, a gem can declare its dependencies without declaring just the name of a dependent package, but the entire identifier (ie: address) of a dependent package.

And, in _just those cases_ where it's necessary to either override a package's location or to complete an incomplete specification of a dependency that is declared only with its name, a fallback and override mechanism like Bundler can be used, if the user choses to do so.

In the end, URIs are just a good, solid way to state a resource, whether with the location or without. And ultimately, it can't _but_ be as applicable to RubyGems as it is to everything else.

I appreciate the concern about resources becoming unavailable, whether due to domain names expiring or any of a whole host of reasons for network resources to become available. But this is a fact of life with network resources and it will never not be a fact of life with network resources.

And the Ruby community has already dealt with such incidences with little fuss and little delay and setback. In the end, the same concern exists for gem that is completely yanked and expunged from its host.

Usually a new version is made available at a new address and we proceed as normal. Arguably that's made easier by Bundler providing an override mechanism. But the point I'm making is that the override mechanism shouldn't have to be part of the default workflow and the default experience.

I appreciate that there are matters of design decisions having been made and time already invested. Nonetheless, I'm providing an alternative take on the design that might be a consideration (or might not).

As for how we use Bundler and RubyGems, in the end I'd just rather not have to use Bundler unless it was strictly necessary. And the only reasons I find myself having to use Bundler is to provide complete dependency specifications for the incomplete dependency declarations inherent to RubyGems and Gemspec. Which means having to declare dependencies twice. And it would be great if that were not necessary, and if some futuristic version of RubyGems leveraged URIs as a standard for both addressing and namespacing.

Thanks for giving the matter some consideration.

**Comment by deivid-rodriguez** [comment]
Thanks @sbellware. I do think this request is reasonable now that I understand it, and worth giving it a thought. There's just too many pressing things to do, and I don't feel this specific feature request will take much priority, specifically since changing the gemspec specification is quite delicate. But I'm leaving this opened for future reevaluation in the mid term.

**Comment by sbellware** [comment]
@deivid-rodriguez 

> There's just too many pressing things to do

No doubt! I appreciate the time you've taken already to consider it and put it in the back of your mind.

**Comment by carlwiedemann** [comment]
We were having quite a few problems with this, as we had private gem `foo` which depended on private gem `bar`, but private gem `bar` happened to have a name conflict with a public gem (yes, the name conflict was an oversight on our part).

The fix from https://github.com/rubygems/rubygems/pull/4609 which went in [2.2.18](https://github.com/rubygems/rubygems/blob/master/bundler/CHANGELOG.md#2218-may-25-2021) did seem to help with this, however it appeared possible that the issue could still be present with older Gemfile.lock files that haven't been reformatted split out the remotes sources into separate sections in the Gemfile.lock.

In other words, if you are able to update bundler to a version `> 2.2.18`, you should also run `bundle update --bundler` to update the Gemfile.lock format if it is not splitting out remote sources.

You may also want to manually update the gem that sources the child private gems -- to to reference my previous example with gem `x` and gem `y`, manually run `bundle update x`.

I'm curious if anyone here is still having this issue after upgrading to a bundler `> 2.2.18`, then running `bundler update --bundler`? @sbellware @vladhadzhiyski @indirect @Fire-Dragon-DoL @bradrobertson

**Comment by synth** [comment]
I'm building a Rails engine which has a dependency on another gem which hasn't published the latest version to rubygems. It is available on Github. How is this supposed to be handled?

**Comment by simi** [comment]
@synth I'm not sure what problem do you face. You can make the gem listed in gemspec dependencies and then on app side you need to specify in Gemfile to use github repo `gem "abc", git: "..."`).

**Comment by synth** [comment]
@simi Thanks for your response. Ok, that sounds like a workaround, but it adds a slight burden to any application that adds the gem. If there are several such dependencies like this, the application needs to individually list them in their Gemfile. This seems to defeat the purpose of adding them as a gemspec dependency in the first place. The purpose of the gemspec dependencies is (in part) to relieve any application that includes that gem from needing to manage said dependencies. 

Having the same gem listed in gemspec as a dependency and listed in including application's Gemfile seems redundant and this redundancy can cause issues. Thanks for your consideration.

**Comment by simi** [comment]
@synth Yes, the key in here is to release the gem to be able to rely on released version.

**Comment by synth** [comment]
@simi This is sadly not always possible for a multitude of reasons. This results in either being blocked or forcing redundancy as I wrote about above.

**Comment by indirect** [comment]
I'm going to try to summarize this issue and the position of the RubyGems team here, for anyone who finds this issue later.

#### Gemspecs do not specify the the source of a dependency

In a gemspec, you can declare a gem name and version, as well as any dependencies of that gem. Dependencies can only be specified by name and version. The source could be RubyGems.org, the source could be a public mirror of RubyGems.org, the source could be a private gem server like Gemstash or Geminabox, or the source could simply be a git repo somewhere. 

#### A gemspec is the wrong place to declare a source

By design, it is not possible to include the _source_ of a gem in a gemspec. We cannot allow gemspecs or gems themselves to hard-code a single specific source, because that would break mirroring, break private gem servers, and break the ability to use a fork of a gem from a git repo.

#### How to declare sources

For dependencies that are not published to RubyGems.org, your users must tell RubyGems and/or Bundler how to reach that source and find that gem. For RubyGems, you use [the `gem source` command](https://guides.rubygems.org/command-reference/#gem-sources). For Bundler, you use the `source` method in the `Gemfile`. Since Bundler exists to configure a single application, it's sources, and its gems, you can use the `Gemfile` and `Gemfile.lock` to require that certain gems or dependencies come from a certain source.

#### RubyGems project policy on sources

As summarized above, the RubyGems team does not believe that any gem should be able to force its dependencies to come from a certain source. Sources should not be controlled by the library author, but by the library user via the Gemfile or `gem` configuration.

#### Future possibilities

Git sources are not supported directly inside RubyGems, but were developed as a feature for Bundler. It is possible that git repo source support might eventually make its way back into RubyGems, bringing `gem install` into parity with other language package managers like `pip` and `npm`. We haven't prioritized this because it's complicated, easily leads to users installing random commits rather than deliberately packaged releases, and seems likely to increase the support burden on both the RubyGems team and the maintainers of any gems installed via this method.

Non-binding "source suggestions" might possibly be implemented as a gemspec feature in the future, allowing gems to suggest a source to check if the configured sources do not contain a dependency. The design of this feature would need to be extremely careful and require manual user opt-in to defend against gems that might try to use source suggestions in a malicious way.

If you're interested in working on something along those lines, feel free to start developing the feature yourself and submit either [an RFC for discussion](https://github.com/rubygems/rfcs) or a draft PR where you can collaborate with the RubyGems team

 **Please note:** Discussing the possibility of these potential features does not guarantee that anything mentioned here will ever be accepted into the codebase, and especially does not include any kind of timeframe.
